Updates to transaction layer doc (#16769)
cockroachdb/docs: v23.1/architecture/transaction-layer.md

### Commits (phase 2)

> CockroachDB checks the running transaction's record to see if it's been `ABORTED`; if it has, it [throws an error](../transaction-retry-error-reference.html) to the client.

Reviewer: nit: Do we want to specify that it is a retriable error here?

> In the common case, it sets the transaction record's state to `STAGING`, and checks the transaction's pending write intents to see if they have succeeded (i.e., been replicated across the cluster).

Reviewer 2: ```suggestion
In the common case, it sets the transaction record's state to `STAGING` and checks the transaction's pending write intents to see if they have been successfully replicated across the cluster.
```
I'm confused about what "In the common case" refers to. I'm also confused about how line 53 relates to line 55 and 57. Who would have set the transaction record to ABORTED? If it's ABORTED, we throw the retryable error and move on. Otherwise, WE set the transaction record's state to STAGING (what else might it have been before?) and check for successful write attempts. When those succeed, then what is the status of the transaction record's state?

Author: This part I did not work on specifically other than add a link, however, I do think you're right that the second paragraph is weirdly phrased. I am adjusting it to clear up some of those odd areas

The timestamp cache is a data structure used to store information about the reads performed by [leaseholders](replication-layer.html#leases). This is used to ensure that once some transaction *t1* reads a row, another transaction *t2* that comes along and tries to write to that row will be ordered after *t1*, thus ensuring a serial order of transactions, aka serializability.

> Whenever a write occurs, its timestamp is checked against the timestamp cache. If the timestamp is less than the timestamp cache's latest value, we attempt to push the timestamp for its transaction forward to a later time. Pushing the timestamp might cause the transaction to restart [during the commit time](#commits-phase-2) of the transaction (see [read refreshing](#read-refreshing)).

Reviewer 2: ```suggestion
Whenever a write occurs, its timestamp is checked against the timestamp cache. If the timestamp is earlier than the timestamp cache's latest value, we attempt to push the timestamp for its transaction forward to a later time. Pushing the timestamp might cause the transaction to restart [during the commit time](#commits-phase-2) of the transaction (see [read refreshing](#read-refreshing)).
```
I don't love the use of "we" in this topic. Who is it? CockroachDB? The client? The end user? It might be better to avoid "we" to be more clear about what the actor is.

"during the commit time of the transaction" reads really strange to me. Do we mean until the transaction is committed?

The timestamp for any transaction, especially a long-running transaction, could be [pushed](#timestamp-cache). An example is when a transaction encounters a key written at a higher timestamp. When this kind of [contention](../performance-best-practices-overview.html#understanding-and-avoiding-transaction-contention) happens between transactions, the closed timestamp **may** provide some mitigation. Increasing the closed timestamp interval may reduce the likelihood that a long-running transaction's timestamp is pushed and must be retried. Thoroughly test any adjustment to the closed timestamp interval before deploying the change in production, because such an adjustment can have an impact on:

- [Follower reads](../follower-reads.html): Latency or throughput may be increased, because reads are served only by the leaseholder.
- [Change data capture](../change-data-capture-overview.html): Latency of changefeed messages may be increased.
- [Statistics collection](../cost-based-optimizer.html#table-statistics): Load placed on the leaseholder may increase during collection.

> While increasing the closed timestamp may decrease retryable errors, it may also increase lock latencies. Consider an example where the long-running transaction `txn 1` holds [write locks](#writing) on keys at time `t=1`, and the long-running transaction `txn 2` is waiting to [read](#reading) those same keys at `t=1`. The following scenarios may occur depending on whether or not the closed timestamp has been increased:

Reviewer 2: ```suggestion
While increasing the closed timestamp may decrease retryable errors, it may also increase lock latencies. Consider an example where:

1. The long-running transaction `txn 1` holds [write locks](#writing) on keys at time `t=1`.
1. The long-running transaction `txn 2` is waiting to [read](#reading) those same keys at `t=1`.

The following scenarios may occur depending on whether or not the closed timestamp has been increased:
```
This is complicated and it might reduce the cognitive load to break it up just a little. WDYT?

> - When the timestamp for `txn 1` is pushed forward by the closed timestamp, moving its writes to time `t=2`, `txn 2` may then be able to proceed and read the keys at `t=1`. The likelihood of retryable errors has increased.

Reviewer 2: ```suggestion
- If the timestamp for `txn 1` is pushed forward by the closed timestamp, its writes are moved to  time `t=2`, and it may then be able to read the keys at `t=1` before writing its changes at `t=2`. The likelihood of retryable errors has increased.
```
Is this correct? I changed it to If to match the next bullet.

> - If the closed timestamp interval is increased, `txn1` may cause `txn2`  to [wait](../performance-recipes.html#waiting-transaction), because `txn 2` cannot read the keys at `t=1` until `txn 1` is complete or is pushed into the future. The likelihood of lock contention has increased.

Reviewer 2: ```suggestion
- If the closed timestamp interval is increased, `txn 2` may need to [wait](../performance-recipes.html#waiting-transaction) for `txn 1` to complete or to be pushed into the future before `txn 2` can proceed. The likelihood of lock contention has increased.
```

Closed timestamps provide the guarantees that are used to provide support for low-latency historical (stale) reads, also known as [Follower Reads](../follower-reads.html). Follower reads can be particularly useful in [multi-region deployments](../multiregion-overview.html).

> When [contention](../performance-best-practices-overview.html#understanding-and-avoiding-transaction-contention) happens between transactions, the closed timestamp can provide some mitigation. Increasing the closed timestamp interval can reduce the likelihood of long-running transactions having their [timestamps pushed](#timestamp-cache), and therefore, causing the transaction to retry. However, note that increasing the closed timestamp interval can have an impact on the following:

Reviewer 3: > and therefore, causing the transaction to retry.

Not always so perhaps loosening the language?

and therefore, reduces the likelihood of the transaction retrying.

> When [contention](../performance-best-practices-overview.html#understanding-and-avoiding-transaction-contention) happens between transactions, the closed timestamp can provide some mitigation. Increasing the closed timestamp interval can reduce the likelihood of long-running transactions having their [timestamps pushed](#timestamp-cache), and therefore, reduces the likelihood of the transaction retrying. However, note that increasing the closed timestamp interval can have an impact on the following:

Reviewer: This is correct, but a double-edged sword.  Considering a scenario in which a long-running transaction named txn1 holds write locks on some keys at time t=1, and txn2 is waiting to read those same keys at t=1:

1. If txn1 has it's timestamp pushed forward by the closed timestamp, moving its writes to t=2, txn2 may be able to proceed and read the keys at t=1.
2. If the closed timestamp interval is raised, that now means that long-running txn1 is causing txn2 to be long-waiting, as it can't read the keys at t=1 until txn1 is done or pushed into the future.

This is in addition to the other callouts listed below, and I think when talking about contention, it's worth mentioning that while increasing the closed timestamp interval may decrease retry errors, in may also increase lock wait latencies.

Closed timestamps provide the guarantees that are used to provide support for low-latency historical (stale) reads, also known as [Follower Reads](../follower-reads.html). Follower reads can be particularly useful in [multi-region deployments](../multiregion-overview.html).

> For long-running transactions, and generally any transaction, it can be inevitable that their timestamp is pushed, such as when they encounter a key written at a higher timestamp. When this kind of [contention](../performance-best-practices-overview.html#understanding-and-avoiding-transaction-contention) happens between transactions, the closed timestamp **may** provide some mitigation. Increasing the closed timestamp interval may reduce the likelihood of long-running transactions having their [timestamps pushed](#timestamp-cache), and therefore, reduces the likelihood of the transaction retrying. However, note that increasing the closed timestamp interval should be **carefully adjusted**, because it can have an impact on the following:

Reviewer 2: ```suggestion
The timestamp for any transaction, especially a long-running transaction, could be [pushed](#timestamp-cache). An example is when a transaction encounters a key written at a higher timestamp. When this kind of [contention](../performance-best-practices-overview.html#understanding-and-avoiding-transaction-contention) happens between transactions, the closed timestamp **may** provide some mitigation. Increasing the closed timestamp interval may reduce the likelihood that a long-running transaction's timestamp is pushed and must be retried. Thoroughly test any adjustment to the closed timestamp interval before deploying the change in production, because such an adjustment can have an impact on:
```
Forgive me if this reworking hurts more than it helps. It's weird to me to link "timestamps pushed" so long after you've introduced the concept. I've also removed some words to try to make the message a bit more clear and less wiggly. WDYT?

> - [Follower reads](../follower-reads.html): Reads will be forced to read from the leaseholder, which may affect latency/throughput.

Reviewer 2: ```suggestion
- [Follower reads](../follower-reads.html): Latency or throughput may be increased, because reads are served only by the leaseholder.
```

Author: Done!

> - [Change data capture](../change-data-capture-overview.html): There could be a lag in changefeed messages.

Reviewer 2: ```suggestion
- [Change data capture](../change-data-capture-overview.html): Latency of changefeed messages may be increased.
```
Latency of ... reading? Writing? Maybe you meant a different kind of "lag".

Author: Done!

Author: Yes, it is latency

> - [Statistics collection](../cost-based-optimizer.html#table-statistics): Collection can place more load on the leaseholder.

Reviewer 2: ```suggestion
- [Statistics collection](../cost-based-optimizer.html#table-statistics): Collection can place more load on the leaseholder.
```
More load when compared with what? than on other nodes? Than before adjusting the setting?

Author: It will cause an increase in load during collection

> While increasing the closed timestamp may decrease retryable errors, it may also increase lock latencies. For example, a long-running transaction (txn 1) holds [write locks](#writing) on keys at time t=1, and txn 2 is waiting to [read](#reading) those same keys at t=1. The following scenarios may occur depending on whether the closed timestamp has been increased or not:

Reviewer 2: ```suggestion
While increasing the closed timestamp may decrease retryable errors, it may also increase lock latencies. Consider an example where the long-running transaction `txn 1` holds [write locks](#writing) on keys at time `t=1`, and the long-running transaction `txn 2` is waiting to [read](#reading) those same keys at `t=1`. The following scenarios may occur, depending on whether or not the closed timestamp has been increased:
```

Author: Changed!

> - If txn 1 has its timestamp pushed forward by the closed timestamp, moving its writes to time t=2, txn 2 may be able to proceed and read the keys at t=1. The chances of retryable errors have increased.

Reviewer: Small nit: it may almost be preferable to write this as "When txn1 has its timestamp pushed", rather than if.

> - When txn 1 has its timestamp pushed forward by the closed timestamp, moving its writes to time t=2, txn 2 may be able to proceed and read the keys at t=1. The chances of retryable errors have increased.

Reviewer 2: ```suggestion
- When the timestamp for `txn 1` is pushed forward by the closed timestamp, to time `t=2`, `txn 2` may then be able to read the keys at `t=1`. The likelihood of retryable errors has increased.
```
There is some subject-predicate disagreement and what seems to me like a dangling modifier. Let me know if I've captured your intention with the edit.

"The chances of...." is really informal, "the chance of" is slightly less informal, but "the likelihood of" is really what you intend here, I think.

Author: Changed, yep, agree!

> - If the closed timestamp interval is raised, long-running txn 1 is making txn 2 [wait](../performance-recipes.html#waiting-transaction), because txn 2 cannot read the keys at t=1 until txn 1 is done or pushed into the future. The chances of lock contention have increased.

Reviewer 2: ```suggestion
- If the closed timestamp interval is increased, `txn1` may cause `txn2`  to [wait](../performance-recipes.html#waiting-transaction), because `txn 2` cannot read the keys at `t=1` until `txn 1` is complete or is pushed into the future. The likelihood of lock contention has increased.
```

Author: Done!

- [Follower reads](../follower-reads.html): Latency or throughput may be increased, because reads are served only by the leaseholder.
- [Change data capture](../change-data-capture-overview.html): Latency of changefeed messages may be increased.
- [Statistics collection](../cost-based-optimizer.html#table-statistics): Load placed on the leaseholder may increase during collection.

While increasing the closed timestamp may decrease retryable errors, it may also increase lock latencies. Consider an example where:

1. The long-running transaction `txn 1` holds [write locks](#writing) on keys at time `t=1`.
1. The long-running transaction `txn 2` is waiting to [read](#reading) those same keys at `t=1`.

The following scenarios may occur depending on whether or not the closed timestamp has been increased:

> - When the timestamp for `txn 1` is pushed forward by the closed timestamp, its writes are moved to  time `t=2`, and it can try again to read the keys at `t=1` before writing its changes at `t=2`. The likelihood of retryable errors has increased.

Reviewer 2: nit: double space in case you have a followup planned.

- Dealing with transactions' state. After a transaction is started, `TxnCoordSender` starts asynchronously sending heartbeat messages to that transaction's transaction record, which signals that it should be kept alive. If the `TxnCoordSender`'s heartbeating stops, the transaction record is moved to the `ABORTED` status.
- Tracking each written key or key range over the course of the transaction.

> - Clearing the accumulated write intent for the transaction when it's committed or aborted. All operations being performed as part of a transaction have to go through the same `TxnCoordSender` to account for all of its write intents, which optimizes the cleanup process.

Reviewer 2: ```suggestion
- Clearing the accumulated write intent for the transaction when it's committed or aborted. All of a transaction's operations go through the same `TxnCoordSender` to account for all of its write intents and optimize the cleanup process.
```
